src/core/algorithms/gfd/egfd_validation.cpp

```diff
@@ -0,0 +1,915 @@
+            typename boost::graph_traits<graph_t>::adjacency_iterator adjacency_it, adjacency_end;
+            boost::tie(adjacency_it, adjacency_end) = boost::adjacent_vertices(vertex, query);
+            for (; adjacency_it != adjacency_end; ++adjacency_it) {
+                if (marked.find(*adjacency_it) == marked.end()) {
+                    marked.insert(*adjacency_it);
+                    next.insert(*adjacency_it);
+                    parent.emplace(*adjacency_it, vertex);
+                }
+            }
+        }
+        current = next;
+    }
+
+    typename boost::graph_traits<graph_t>::edge_iterator it_edge, end_edge;
+    for (boost::tie(it_edge, end_edge) = edges(query); it_edge != end_edge; ++it_edge) {
+        vertex_t origin = boost::source(*it_edge, query);
+        vertex_t finish = boost::target(*it_edge, query);
+        if ((parent.find(origin) != parent.end()) && (parent.find(finish) != parent.end()) &&
+            (parent.at(origin) != finish) && (parent.at(finish) != origin)) {
+            int origin_level = 0;
+            int finish_level = 0;
+            for (std::size_t i = 0; i < levels.size(); ++i) {
+                if (levels.at(i).find(origin) != levels.at(i).end()) {
+                    origin_level = i;
+                }
+                if (levels.at(i).find(finish) != levels.at(i).end()) {
+                    finish_level = i;
+                }
+            }
+            if (origin_level == finish_level) {
+                snte.insert(*it_edge);
+            }
+            nte.insert(*it_edge);
+        }
+    }
+}
+
+void EGfdValidation::TopDownConstruct(CPI& cpi, const graph_t& graph, const graph_t& query,
+                                      const std::vector<std::set<vertex_t>>& levels,
+                                      const std::map<vertex_t, vertex_t>& parent,
+                                      std::map<vertex_t, std::set<vertex_t>>& candidates,
+                                      const std::set<edge_t>& snte) {
+    vertex_t root = *levels.at(0).begin();
+    typename boost::graph_traits<graph_t>::vertex_iterator it, end;
+    for (boost::tie(it, end) = vertices(query); it != end; ++it) {
+        std::set<vertex_t> empty = {};
+        candidates.emplace(*it, empty);
+    }
+
+    for (boost::tie(it, end) = vertices(graph); it != end; ++it) {
+        if (graph[*it].attributes.at("label") == query[root].attributes.at("label") &&
+            boost::degree(*it, graph) >= boost::degree(root, query) &&
+            CandVerify(graph, *it, query, root)) {
+            candidates.at(root).insert(*it);
+        }
+    }
+    std::set<vertex_t> visited = {root};
+    std::map<vertex_t, std::set<vertex_t>> visited_neighbours;
+    std::map<vertex_t, std::set<vertex_t>> unvisited_neighbours;
+    std::map<vertex_t, int> cnts;
+
+    std::vector<std::set<vertex_t>>::const_iterator i = std::next(levels.cbegin());
+    for (; i != levels.cend(); ++i) {

```

Comment:
this huge loop should be decomposed, you can:
- split body into separate parts
- extract whole body into separate method